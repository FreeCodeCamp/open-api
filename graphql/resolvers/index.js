import { mergeResolvers } from 'merge-graphql-schemas';
import { userResolvers } from './user';

export { directiveResolvers } from './directives';
export default mergeResolvers([userResolvers]);
