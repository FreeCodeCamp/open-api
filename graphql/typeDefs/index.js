import { mergeTypes } from 'merge-graphql-schemas';
import User from './User';
import directives from './directives';

export default mergeTypes([User, directives]);
