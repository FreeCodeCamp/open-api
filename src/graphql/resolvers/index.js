import { mergeResolvers } from 'merge-graphql-schemas';
import { userResolvers } from './user';

export { createDirectives } from './directives';
export default mergeResolvers([userResolvers]);
