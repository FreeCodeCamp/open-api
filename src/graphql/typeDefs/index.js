import { mergeTypes } from 'merge-graphql-schemas';
import User from './User';
import GDPRUser from './GDPR-User';
import directives from './directives';

export default mergeTypes([User, GDPRUser, directives]);
