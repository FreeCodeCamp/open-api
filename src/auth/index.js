import jwt from 'jsonwebtoken';
import debug from 'debug';

import { AuthorizationError } from '../graphql/errors';

const log = debug('fcc:auth');
const { JWT_CERT } = process.env;

export const namespace = 'https://www.freecodecamp.org/';

export const getTokenFromContext = ctx =>
  ctx && ctx.headers && ctx.headers.authorization;

export function verifyWebToken(ctx) {
  log('Verifying token');
  const token = getTokenFromContext(ctx);
  if (!token) {
    throw new AuthorizationError({
      message:
        'You must supply a JSON Web Token for authorization, are you logged in?'
    });
  }
  let decoded = null;
  let error = null;
  try {
    decoded = jwt.verify(token.replace('Bearer ', ''), JWT_CERT);
  } catch (err) {
    error = err;
  } finally {
    return { decoded, error, isAuth: !!decoded };
  }
}
