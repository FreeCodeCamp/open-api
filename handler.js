import { graphqlLambda, graphiqlLambda } from 'apollo-server-lambda';
import lambdaPlayground from 'graphql-playground-middleware-lambda';
import { makeExecutableSchema } from 'graphql-tools';
import debug from 'debug';

import typeDefs from './graphql/typeDefs';
import { default as resolvers, directiveResolvers } from './graphql/resolvers';

const log = debug('fcc:handler');

export const graphqlSchema = makeExecutableSchema({
  typeDefs,
  resolvers,
  directiveResolvers,
  logger: console
});

// Database connection logic lives outside of the handler for performance reasons
const connectToDatabase = require('./dataLayer');

const server = require('apollo-server-lambda');

export async function graphqlHandler(event, context, callback) {
  /* Cause Lambda to freeze the process and save state data after
  the callback is called the effect is that new handler invocations
  will be able to re-use the database connection.
  See https://docs.aws.amazon.com/lambda/latest/dg/nodejs-prog-model-context.html
  and https://www.mongodb.com/blog/post/optimizing-aws-lambda-performance-with-mongodb-atlas-and-nodejs */
  context.callbackWaitsForEmptyEventLoop = false;

  function callbackFilter(error, output) {
    if (!output.headers) {
      output.headers = {};
    }
    // eslint-disable-next-line no-param-reassign
    output.headers['Access-Control-Allow-Origin'] = '*';
    output.headers['Access-Control-Allow-Credentials'] = true;
    output.headers['Content-Type'] = 'application/json';

    callback(error, output);
  }

  const handler = server.graphqlLambda((event, context) => {
    const { headers } = event;
    const { functionName } = context;

    return {
      schema: graphqlSchema,
      context: {
        headers,
        functionName,
        event,
        context
      }
    };
  });

  try {
    await connectToDatabase();
  } catch (err) {
    log('MongoDB connection error: ', err);
    // TODO: return 500?
    process.exit();
  }
  return handler(event, context, callbackFilter);
}

export const apiHandler = lambdaPlayground({
  endpoint: process.env.GRAPHQL_ENDPOINT_URL
    ? process.env.GRAPHQL_ENDPOINT_URL
    : '/production/graphql'
});
