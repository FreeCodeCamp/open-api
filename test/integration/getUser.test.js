/* global beforeAll afterAll expect */
import mongoose from 'mongoose';
import { graphql } from 'graphql';

import { graphqlSchema } from '../../src/handler';
import UserModel from '../../src/dataLayer/model/user';

beforeAll(async function beforeAllTests() {
  await mongoose.connect(global.__MONGO_URI__);
});

afterAll(async function afterAllTests() {
  await mongoose.disconnect();
});

const contextNoToken = global.mockedContextWithOutToken;
const invalidContext = global.mockedContextWithInValidToken;
const validContext = global.mockedContextWithValidToken;

// language=GraphQL
const query = `
    query {
        getUser(email: "user@example.com") {
            name
            email
        }
    }
  `;

const expectedNoUserQuery = `
    query {
      getUser(email: "nowhere@tobe.seen") {
        name
        email
      }
    }
  `;

const malformedQuery = `
    query {
      getUser(email: "Ooops!") {
        name
        email
      }
    }
  `;

const rootValue = {};
describe('getUser', () => {
  it('should return null and an auth error without a token', done => {
    expect.assertions(2);

    graphql(graphqlSchema, query, rootValue, contextNoToken)
      .then(({ data, errors }) => {
        expect(data.getUser).toMatchSnapshot('no user');
        expect(errors).toMatchSnapshot('not logged in');
        return;
      })
      .then(done)
      .catch(done);
  });

  it('should return null and an auth error with an invalid token', done => {
    expect.assertions(2);

    graphql(graphqlSchema, query, rootValue, invalidContext)
      .then(({ data, errors }) => {
        expect(data.getUser).toMatchSnapshot('no user');
        expect(errors).toMatchSnapshot('no auth error');
        return;
      })
      .then(done)
      .catch(done);
  });

  it('should return a user after one has been created', done => {
    expect.assertions(2);

    const user = new UserModel({
      name: 'user',
      email: 'user@example.com',
      accountLinkId: 'a-unique-string-for-user@example.com'
    });

    user.save().then(() => {
      graphql(graphqlSchema, query, rootValue, validContext)
        .then(({ data, errors }) => {
          expect(data).toMatchSnapshot('user found');
          expect(errors).toMatchSnapshot('no errors');
          return;
        })
        .then(done)
        .catch(done);
    });
  });

  it('should return null if no user has been found', done => {
    expect.assertions(2);

    graphql(graphqlSchema, expectedNoUserQuery, rootValue, validContext)
      .then(result => {
        const { data, errors } = result;
        expect(data.getUser).toBe(null);
        expect(errors).toMatchSnapshot('no user error');
        return;
      })
      .then(done)
      .catch(done);
  });

  it('should return errors for a malformed query', done => {
    expect.assertions(2);

    graphql(graphqlSchema, malformedQuery, rootValue, validContext)
      .then(result => {
        const { data, errors } = result;
        expect(data.getUser).toBe(null);
        expect(errors).toMatchSnapshot('malformed query error');
        return;
      })
      .then(done)
      .catch(done);
  });
});
