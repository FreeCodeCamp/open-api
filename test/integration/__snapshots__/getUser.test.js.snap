// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getUser should return a user after one has been created: no errors 1`] = `undefined`;

exports[`getUser should return a user after one has been created: user found 1`] = `
Object {
  "getUser": Object {
    "email": "user@example.com",
    "name": "user",
  },
}
`;

exports[`getUser should return errors for a malformed query: malformed query error 1`] = `
Array [
  [GraphQLError: Something went wrong, please check and try again],
]
`;

exports[`getUser should return null and an auth error with an invalid token: no auth error 1`] = `
Array [
  [GraphQLError: You are not authorized, jwt malformed],
]
`;

exports[`getUser should return null and an auth error with an invalid token: no user 1`] = `null`;

exports[`getUser should return null and an auth error without a token: no user 1`] = `null`;

exports[`getUser should return null and an auth error without a token: not logged in 1`] = `
Array [
  [GraphQLError: You must supply a JSON Web Token for authorization, are you logged in?],
]
`;

exports[`getUser should return null if no user has been found: no user error 1`] = `
Array [
  [GraphQLError: Something went wrong, please check and try again],
]
`;
